Local cartesian conversion utility

aaron-bray opened the pull request:
Based on the NGA GeoTrans Library

mleotta commented on vital/types/local_cartesian.h:
```diff
@@ -0,0 +1,106 @@
+   * @param [in] orientation     : Orientation angle of the local cartesian coordinate system,
+   *                               in radians along the Z axis, normal to the earth surface
+   */
+  void set_origin( geo_point const& origin, double orientation=0 );
```
Should orientation be a full 3D rotation?  If we are going to add orientation, why limit it to the Z axis?  I'm not saying this has to change, I'm just curious about the motivation for this.

aaron-bray replied:
I am not sure the motivation for this.
This is pretty old NGA code that I am just porting into kwiver.
I never used the orientation when I used this, but the code does support it if needed.
Not sure what use case they were coding for

mleotta replied:
that's fine, we can leave it as is.

mleotta commented on vital/types/local_cartesian.cxx:
```diff
@@ -0,0 +1,524 @@
+const double TWO_PI = (2.0 * PI);
+const double DEG_TO_RAD = (PI / 180.0);
+const double RAD_TO_DEG = (180.0 / PI);
+const double AD_C = 1.0026000;            /* Toms region 1 constant */
```
These last two, `AD_C` and `COS_67P5` are custom for this code and probably don't belong in `math_constants.h`.  However, they are only used in one function, so I would just make them local to that function.

linus-sherrill replied:
These could be `constexpr` values.

mleotta commented on vital/types/local_cartesian.cxx:
```diff
@@ -0,0 +1,524 @@
+    double Z = geocentric_coordinate.z();
+    double latitude, longitude, height;
+
+    if (Iterative_Geocent_algorithm)
```
What does this compile-time switch do?  Should we just remove the code that is not being used?  Or maybe make it a run time option?  It looks like there are two algorithm alternatives, do we know the pros and cons of each?

aaron-bray replied:
Well, it looks like they released geotrans 3.8 this year, but I need certs to download it...
It looks like they added spherical coordinate support and fixed a but we may want to get (if I can get the source code)

https://earth-info.nga.mil/GandG/update/wgs84/apps/geotrans/current-version/docs/MSP_Geotrans_Release_Notes%203.8.pdf

mleotta commented on vital/tests/test_local_cartesian.cxx:
```diff
@@ -0,0 +1,129 @@
+
+  // Get the geopoint from an offset
+  lc_lla.convert_from_cartesian(offset1, geo_out);
+  auto const epsilon_geo_out_to_geo1 = (geo_out.location() - geo1.location()).norm();
```
consider `lpNorm<Infinity>()` instead of `norm()`.  That gives you the maximum absolution value of the elements.  The L2 distance doesn't make as much sense here.  Also, you really just want to make sure the error is low in all elements, so checking the max is probably the right thing to do.

mleotta replied:
here are the docs:
https://eigen.tuxfamily.org/dox/group__TutorialReductionsVisitorsBroadcasting.html

I would not use `norm()` here because the units on altitude are very different from those on lat and long

aaron-bray replied:
I just switched the test to test each component individually, seems more flexible and straight forward

mleotta commented on vital/tests/test_local_cartesian.cxx:
```diff
@@ -0,0 +1,129 @@
+  // Get the geopoint from an offset
+  lc_lla.convert_from_cartesian(offset1, geo_out);
+  auto const epsilon_geo_out_to_geo1 = (geo_out.location() - geo1.location()).norm();
+  EXPECT_MATRIX_NEAR(geo_out.location(), geo1.location(), 1e-3);
```
This error tolerance seems quite high.  I think 1e-3 in lat/long is about 100 meters.  Is this conversion not more accurate than that?

aaron-bray replied:
Note the matrix compare eigen is doing is not based on the decimal places (not sure how it does this), but you can see the accuracy of the conversions are pretty good down to at least the 6th decimal place

geo_out.location() evaluates to   -73.758985147971799
   42.850126094497369
0.0002865474671125412
geo1.location() evaluates to -73.758985150000001
 42.850126090000003
                  0
1e-4 evaluates to 0.0001

mleotta replied:
The latitude and longitude values need to be a lot more precise than altitude.  I would not score these against the same epsilon.  1e-4 is sub-millimeter in height but 10s of meters in horizontal position.

aaron-bray replied:
I just switched the test to test each component individually, seems more flexible and straight forward

aaron-bray commented:
@mleotta  I Think this is ready to go. I updated the test to compare individual components, and removed the legacy geocentric-to-geodetic algorithm. I did not update the code styling just so the code base is easily compatible to the original geotrans code just in case. I originally did a few changes to get things into a class with better functions, but for the most part, I tried to preserve styling to the original.
